File: Day4/PerformanceComp3/cpp_project/prime_benchmark.cpp
#include <iostream>
#include <vector>
#include <chrono>

int main() {
    const int LIMIT = 100'000'000;
    std::cout << "Calculating primes up to " << LIMIT << "..." << std::endl;

    auto start_time = std::chrono::high_resolution_clock::now();

    // std::vector<bool> is space-optimized (1 bit per element) in C++
    std::vector<bool> is_prime(LIMIT + 1, true);
    is_prime[0] = false;
    is_prime[1] = false;

    for (int p = 2; p * p <= LIMIT; p++) {
        if (is_prime[p]) {
            for (int i = p * p; i <= LIMIT; i += p) {
                is_prime[i] = false;
            }
        }
    }

    int prime_count = 0;
    for (int p = 2; p <= LIMIT; p++) {
        if (is_prime[p]) {
            prime_count++;
        }
    }

    auto end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> diff = end_time - start_time;

    std::cout << "Sieve complete!" << std::endl;
    std::cout << "Primes found: " << prime_count << std::endl;
    std::cout << "C++ Prime Sieve Time: " << diff.count() << " seconds" << std::endl;

    return 0;
}

File: Day4/PerformanceComp3/rust_project/src/bin/prime_benchmark.rs
use std::time::Instant;

fn main() {
    let limit = 100_000_000;
    println!("Calculating primes up to {}...", limit);

    let start_time = Instant::now();

    // Vec<bool> in Rust is byte-aligned (1 byte per bool)
    // This uses more memory than C++'s vector<bool> but is significantly faster to access
    let mut is_prime = vec![true; limit + 1];
    is_prime[0] = false;
    is_prime[1] = false;

    let mut p = 2;
    while p * p <= limit {
        if is_prime[p] {
            let mut i = p * p;
            while i <= limit {
                is_prime[i] = false;
                i += p;
            }
        }
        p += 1;
    }

    let mut prime_count = 0;
    for p in 2..=limit {
        if is_prime[p] {
            prime_count += 1;
        }
    }

    let duration = start_time.elapsed();

    println!("Sieve complete!");
    println!("Primes found: {}", prime_count);
    println!("Rust Prime Sieve Time: {:.4} seconds", duration.as_secs_f64());
}
